### ISSUE TITLE
Broken app label: microsoftvisualstudiocode

### ISSUE THREAD

**Issue by nohartma** [Installomator/Installomator#1963]
ERROR: Error downloading https://go.microsoft.com/fwlink/?LinkID=2156837

Installomator Version 10.6

Below you can find the log:

[STEP 1 of 5]
Executing Policy Visual Studio Code
[STEP 2 of 5]
Running script installomator.10.6.sh...
Script exit code: 2
Script result: 2024-10-17 14:07:14 : REQ   :  : shifting arguments for Jamf
2024-10-17 14:07:14 : REQ   : visualstudiocode : ################## Start Installomator v. 10.7beta, date 2024-09-02
2024-10-17 14:07:14 : INFO  : visualstudiocode : ################## Version: 10.7beta
2024-10-17 14:07:14 : INFO  : visualstudiocode : ################## Date: 2024-09-02
2024-10-17 14:07:14 : INFO  : visualstudiocode : ################## visualstudiocode
2024-10-17 14:07:14 : DEBUG : visualstudiocode : DEBUG mode 1 enabled.
2024-10-17 14:07:14 : INFO  : visualstudiocode : setting variable from argument DEBUG=0
2024-10-17 14:07:14 : INFO  : visualstudiocode : setting variable from argument NOTIFY=silent
2024-10-17 14:07:14 : INFO  : visualstudiocode : setting variable from argument BLOCKING_PROCESS_ACTION=kill
2024-10-17 14:07:14 : INFO  : visualstudiocode : setting variable from argument INSTALL=""
2024-10-17 14:07:14 : INFO  : visualstudiocode : setting variable from argument NOTIFY_DIALOG=1
2024-10-17 14:07:14 : DEBUG : visualstudiocode : name=Visual Studio Code
2024-10-17 14:07:14 : DEBUG : visualstudiocode : appName=Visual Studio Code.app
2024-10-17 14:07:14 : DEBUG : visualstudiocode : type=zip
2024-10-17 14:07:14 : DEBUG : visualstudiocode : archiveName=
2024-10-17 14:07:14 : DEBUG : visualstudiocode : downloadURL=https://go.microsoft.com/fwlink/?LinkID=2156837
2024-10-17 14:07:14 : DEBUG : visualstudiocode : curlOptions=
2024-10-17 14:07:14 : DEBUG : visualstudiocode : appNewVersion=
2024-10-17 14:07:14 : DEBUG : visualstudiocode : appCustomVersion function: Not defined
2024-10-17 14:07:14 : DEBUG : visualstudiocode : versionKey=CFBundleShortVersionString
2024-10-17 14:07:14 : DEBUG : visualstudiocode : packageID=
2024-10-17 14:07:14 : DEBUG : visualstudiocode : pkgName=
2024-10-17 14:07:14 : DEBUG : visualstudiocode : choiceChangesXML=
2024-10-17 14:07:14 : DEBUG : visualstudiocode : expectedTeamID=UBF8T346G9
2024-10-17 14:07:14 : DEBUG : visualstudiocode : blockingProcesses=Code
2024-10-17 14:07:14 : DEBUG : visualstudiocode : installerTool=
2024-10-17 14:07:15 : DEBUG : visualstudiocode : CLIInstaller=
2024-10-17 14:07:15 : DEBUG : visualstudiocode : CLIArguments=
2024-10-17 14:07:15 : DEBUG : visualstudiocode : updateTool=
2024-10-17 14:07:15 : DEBUG : visualstudiocode : updateToolArguments=
2024-10-17 14:07:15 : DEBUG : visualstudiocode : updateToolRunAsCurrentUser=
2024-10-17 14:07:15 : INFO  : visualstudiocode : BLOCKING_PROCESS_ACTION=kill
2024-10-17 14:07:15 : INFO  : visualstudiocode : NOTIFY=silent
2024-10-17 14:07:15 : INFO  : visualstudiocode : LOGGING=DEBUG
2024-10-17 14:07:15 : INFO  : visualstudiocode : LOGO=/System/Applications/App Store.app/Contents/Resources/AppIcon.icns
2024-10-17 14:07:15 : INFO  : visualstudiocode : Label type: zip
2024-10-17 14:07:15 : INFO  : visualstudiocode : archiveName: Visual Studio Code.zip
2024-10-17 14:07:15 : DEBUG : visualstudiocode : Changing directory to /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/tmp.LLn9asqb0y
2024-10-17 14:07:15 : INFO  : visualstudiocode : name: Visual Studio Code, appName: Visual Studio Code.app
2024-10-17 14:07:15.185 mdfind[6552:147350] [UserQueryParser] Loading keywords and predicates for locale "en_US"
2024-10-17 14:07:15.186 mdfind[6552:147350] [UserQueryParser] Loading keywords and predicates for locale "en"
2024-10-17 14:07:15.290 mdfind[6552:147350] Couldn't determine the mapping between prefab keywords and predicates.
2024-10-17 14:07:15 : WARN  : visualstudiocode : No previous app found
2024-10-17 14:07:15 : WARN  : visualstudiocode : could not find Visual Studio Code.app
2024-10-17 14:07:15 : INFO  : visualstudiocode : appversion: 
2024-10-17 14:07:15 : INFO  : visualstudiocode : Latest version not specified.
2024-10-17 14:07:15 : REQ   : visualstudiocode : Downloading https://go.microsoft.com/fwlink/?LinkID=2156837 to Visual Studio Code.zip
2024-10-17 14:07:15 : DEBUG : visualstudiocode : No Dialog connection, just download
2024-10-17 14:07:16 : ERROR : visualstudiocode : error downloading https://go.microsoft.com/fwlink/?LinkID=2156837
ls: Visual Studio Code.zip: No such file or directory
2024-10-17 14:07:16 : ERROR : visualstudiocode : File list: 
2024-10-17 14:07:16 : ERROR : visualstudiocode : File type: Visual Studio Code.zip: cannot open `Visual Studio Code.zip' (No such file or directory)
2024-10-17 14:07:16 : DEBUG : visualstudiocode : Deleting /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/tmp.LLn9asqb0y
2024-10-17 14:07:16 : DEBUG : visualstudiocode : Debugging enabled, Deleting tmpDir output was:
/var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/tmp.LLn9asqb0y
2024-10-17 14:07:16 : INFO  : visualstudiocode : Installomator did not close any apps, so no need to reopen any apps.
2024-10-17 14:07:16 : ERROR : visualstudiocode : ERROR: Error downloading https://go.microsoft.com/fwlink/?LinkID=2156837 error:
* Host go.microsoft.com:443 was resolved.
* IPv6: (none)
* IPv4: 23.32.54.57
*   Trying 23.32.54.57:443...
* Connected to go.microsoft.com (23.32.54.57) port 443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
} [321 bytes data]
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* (304) (IN), TLS handshake, Unknown (8):
{ [35 bytes data]
* (304) (IN), TLS handshake, Certificate (11):
{ [3686 bytes data]
* (304) (IN), TLS handshake, CERT verify (15):
{ [264 bytes data]
* (304) (IN), TLS handshake, Finished (20):
{ [36 bytes data]
* (304) (OUT), TLS handshake, Finished (20):
} [36 bytes data]
* SSL connection using TLSv1.3 / AEAD-CHACHA20-POLY1305-SHA256 / [blank] / UNDEF
* ALPN: server accepted http/1.1
* Server certificate:
*  subject: C=US; ST=WA; L=Redmond; O=Microsoft Corporation; CN=go.microsoft.com
*  start date: Aug 29 20:17:58 2024 GMT
*  expire date: Aug 24 20:17:58 2025 GMT
*  subjectAltName: host "go.microsoft.com" matched cert's "go.microsoft.com"
*  issuer: C=US; O=Microsoft Corporation; CN=Microsoft Azure RSA TLS Issuing CA 07
*  SSL certificate verify ok.
* using HTTP/1.x
> GET /fwlink/?LinkID=2156837 HTTP/1.1

> Host: go.microsoft.com

> User-Agent: curl/8.7.1

> Accept: */*

> 

* Request completely sent off
< HTTP/1.1 302 Moved Temporarily

< Content-Length: 0

< Server: Kestrel

< Location: https://code.visualstudio.com/sha/download?build=stable&os=darwin-universal

< Request-Context: appId=cid-v1:9b037ab9-fa5a-4c09-81bd-41ffa859f01e

< X-Response-Cache-Status: True

< Expires: Thu, 17 Oct 2024 12:07:15 GMT

< Cache-Control: max-age=0, no-cache, no-store

< Pragma: no-cache

< Date: Thu, 17 Oct 2024 12:07:15 GMT

< Connection: keep-alive

< Strict-Transport-Security: max-age=31536000 ; includeSubDomains

< 

* Ignoring the response-body
* Connection #0 to host go.microsoft.com left intact
* Issue another request to this URL: 'https://code.visualstudio.com/sha/download?build=stable&os=darwin-universal'
* Host code.visualstudio.com:443 was resolved.
* IPv6: (none)
* IPv4: 13.107.246.45
*   Trying 13.107.246.45:443...
* Connected to code.visualstudio.com (13.107.246.45) port 443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
} [326 bytes data]
* (304) (IN), TLS handshake, Server hello (2):
{ [88 bytes data]
* (304) (OUT), TLS handshake, Client hello (1):
} [359 bytes data]
* (304) (IN), TLS handshake, Server hello (2):
{ [155 bytes data]
* (304) (IN), TLS handshake, Unknown (8):
{ [19 bytes data]
* (304) (IN), TLS handshake, Certificate (11):
{ [4555 bytes data]
* (304) (IN), TLS handshake, CERT verify (15):
{ [264 bytes data]
* (304) (IN), TLS handshake, Finished (20):
{ [52 bytes data]
* (304) (OUT), TLS handshake, Finished (20):
} [52 bytes data]
* SSL connection using TLSv1.3 / AEAD-AES256-GCM-SHA384 / [blank] / UNDEF
* ALPN: server accepted h2
* Server certificate:
*  subject: C=US; ST=WA; L=Redmond; O=Microsoft Corporation; CN=code.visualstudio.com
*  start date: Aug 12 16:25:17 2024 GMT
*  expire date: Feb  8 16:25:17 2025 GMT
*  subjectAltName: host "code.visualstudio.com" matched cert's "code.visualstudio.com"
*  issuer: C=US; O=Microsoft Corporation; CN=Microsoft Azure RSA TLS Issuing CA 04
*  SSL certificate verify ok.
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://code.visualstudio.com/sha/download?build=stable&os=darwin-universal
* [HTTP/2] [1] [:method: GET]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: code.visualstudio.com]
* [HTTP/2] [1] [:path: /sha/download?build=stable&os=darwin-universal]
* [HTTP/2] [1] [user-agent: curl/8.7.1]
* [HTTP/2] [1] [accept: */*]
> GET /sha/download?build=stable&os=darwin-universal HTTP/2

> Host: code.visualstudio.com

> User-Agent: curl/8.7.1

> Accept: */*

> 

* Request completely sent off
< HTTP/2 403 

< date: Thu, 17 Oct 2024 12:07:16 GMT

< content-type: text/html

< content-length: 1261

< cache-control: no-store

< x-azure-ref: 20241017T120716Z-176d4fdd79cdfrzg8f74fxntqn0000000cn000000000gmm5

< x-cache: CONFIG_NOCACHE

< 

* The requested URL returned error: 403
* process_pending_input: nghttp2_session_mem_recv() returned -902:The user callback function failed
* Connection #1 to host code.visualstudio.com left intact
curl: (56) The requested URL returned error: 403

2024-10-17 14:07:16 : REQ   : visualstudiocode : ################## End Installomator, exit code 2 
Error running script: return code was 2.
[STEP 3 of 5]
[STEP 4 of 5]
Inventory will be updated when all queued actions in Self Service are complete.
[STEP 5 of 5]

**Comment by kernsb** [comment]
As far as I can tell, this could be fixed with the following:

```
appNewVersion=$( \
  curl -fsL -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0.1 Safari/605.1.15' "https://code.visualstudio.com/Updates" | \
  grep "/darwin" | \
  grep -oiE ".com/([^>]+)([^<]+)/darwin" | \
  cut -d "/" -f 2 | \
  sed $'s/[^[:print:]	]//g' | \
  head -1 \
)

downloadURL="https://update.code.visualstudio.com/${appNewVersion}/darwin-universal/stable" # Universal
```

Issues in the label that I found were:

- New version wasn't being found because the check URL redirect wasn't being followed without having a User Agent header specified.

- Once we had the new version being found again, we can construct the URL directly instead of relying on a redirect.

**Comment by kernsb** [comment]
Fixed in #1964
